### PR TITLE
Migrate to newer DD4hep revision

### PR DESCRIPTION
**Pull request by zaborowska**
Following the changes in DD4hep (>November 2015) that accompanied their migration to Geant multithreaded, a new class for the Geant detector construction is needed to properly translate the detector to G4 together with the sensitive detectors.
It invokes an automatic translation of the detector to Geant4 (from DD4hep::Simulation::Geant4DetectorConstruction::Construct() ) and also includes the creation of SD (from DD4hep::Simulation::Geant4DetectorSensitivesConstruction::constructSensitives).
(TODO: perhaphs a better solution can be found, without code repetition?)

Additionally, sensitive detectors are created: both SimpleTrackerSD and SimpleCalorimeterSD that mimic the internal DD4hep SD classes (Geant4Tracker, Geant4Calorimeter).
Class DetectorDescription/DetSensitive/src/SDWrapper.cpp creates DD4hep factory methods to construct user-defined SDs.
TODO: documentation

Moreover, Test/TestDD4hep directory is created with a simple box geometry to test segmentation, SDs, hits number, agreement of internal DD4hep SD and user-defined SD etc...
Also, it includes a service that allows dumping the geometry to GDML (for instance: to compare against the standalone G4 simulation)

### CODE REVIEW

**Comment by hegner** [comment]
test this please

**Comment by jlingema** [comment]
Hi Anna,
I had a quick look at your changes: Could you re-use the `elements.xml` and `materials.xml` from DetectorDescription?
Cheers,
Joschka

**Comment by zaborowska** [comment]
Hi Joschka,
I removed those xmls.

**Comment by zaborowska** [comment]
As discussed with Jana, FCC-hh ECal cell size was increased to x/y=0.5 mm.
I added more comments to the files as well.

**Comment by jlingema** [comment]
retest this please

**Comment by jlingema** [comment]
retest this please

**Comment by jlingema** [comment]
retest this please

**Comment by jlingema** [comment]
retest this please

File: DetectorDescription/DetDesServices/src/GeoConstruction.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+  typedef DD4hep::Geometry::GeoHandlerTypes::ConstVolumeSet VolSet;
+  DD4hep::Simulation::Geant4GeometryInfo* p = DD4hep::Simulation::Geant4Mapping::instance().ptr();
+  _SV& vols = p->sensitives;
+
+  for(_SV::const_iterator iv=vols.begin(); iv != vols.end(); ++iv)  {
+    DD4hep::Geometry::SensitiveDetector sd = (*iv).first;
+    std::string typ = sd.type(), nam = sd.name();
+    G4VSensitiveDetector* g4sd = 
+      DD4hep::PluginService::Create<G4VSensitiveDetector*>(typ, nam, &m_lcdd);
+    if (!g4sd) {
```
Comment (hegner) [inline DetectorDescription/DetDesServices/src/GeoConstruction.cpp]: nullptr comparisons are cleaner code

File: DetectorDescription/DetDesServices/src/GeoConstruction.h [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+namespace det {
+class GeoConstruction : public G4VUserDetectorConstruction {
+public:
+  /// Constructor
+  GeoConstruction(DD4hep::Geometry::LCDD& lcdd);
+  /// Default destructor
+  virtual ~GeoConstruction();
+  /// Geometry construction callback: Invoke the conversion to Geant4
+  virtual G4VPhysicalVolume* Construct() final;
```
Comment (hegner) [inline DetectorDescription/DetDesServices/src/GeoConstruction.h]: who owns it in the end and has to destruct it?

File: DetectorDescription/DetSensitive/DetSensitive/SimpleTrackerSD.h [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+   *  @param aHitsCollections Geant hits collection.
+   */
+  virtual void Initialize(G4HCofThisEvent* aHitsCollections) final;
+  /** Process hit once the particle hit the sensitive volume.
+   *  Checks if the energy deposit is larger than 0, calculates the position and cellID,
+   *  saves that into the hit collection.
+   *  New hit is created for each energy deposit (to save information about time)
+   *  @param aStep Step in which particle deposited the energy.
+   */
+  virtual bool ProcessHits(G4Step*aStep, G4TouchableHistory*) final;
```
Comment (hegner) [inline DetectorDescription/DetSensitive/DetSensitive/SimpleTrackerSD.h]: missing space ;-)

File: DetectorDescription/DetSensitive/src/SDWrapper.cpp [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+#include "DD4hep/LCDD.h"
+
+#include "DetSensitive/SimpleTrackerSD.h"
+#include "DetSensitive/SimpleCalorimeterSD.h"
+
+namespace DD4hep {
+namespace Simulation {
+
+// Factory method to create an instance of SimpleTrackerSD
+static G4VSensitiveDetector* create_simple_tracker_sd(
```
Comment (hegner) [inline DetectorDescription/DetSensitive/src/SDWrapper.cpp]: Ownership goes to the caller? Same question for the function further below. Just add a line to the comment.

File: DetectorDescription/DetSensitive/src/SimpleCalorimeterSD.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+#include "CLHEP/Vector/ThreeVector.h"
+
+namespace det {
+SimpleCalorimeterSD::SimpleCalorimeterSD(const std::string& aDetectorName,
+  const std::string& aReadoutName,
+  const DD4hep::Geometry::Segmentation& aSeg)
+  : G4VSensitiveDetector(aDetectorName), m_seg(aSeg) {
+  // name of the collection of hits is determined byt the readout name (from XML)
+  collectionName.insert(aReadoutName);
+  std::cout<<" Adding a collection with the name: "<<aReadoutName<<std::endl;
```
Comment (hegner) [inline DetectorDescription/DetSensitive/src/SimpleCalorimeterSD.cpp]: no unconditional cout please. Use the MsgSvc.

File: DetectorDescription/DetSensitive/src/SimpleTrackerSD.cpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    track->GetTrackID(), track->GetDefinition()->GetPDGEncoding(),edep, track->GetGlobalTime());
+  if ( hit )  {
+    hit->cellID  = segmentation::cellID(m_seg, *aStep);
+    hit->energyDeposit = edep;
+    hit->position = position;
+    hit->momentum = direction;
+    trackerCollection->insert(hit);
+    return true;
+  }
+  return false;
```
Comment (hegner) [inline DetectorDescription/DetSensitive/src/SimpleTrackerSD.cpp]: could you add a line about in which circumstances you end up here? and why that is OK?

File: DetectorDescription/DetSensitive/tests/scripts/cmpTrackerSdDetailed.py [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    return retrieve(cellId, 0b000000111111000000,1*bitfieldsize)
+
+def x(cellId):
+    return retrieve(cellId, 0b111111000000000000,2*bitfieldsize)
+
+def cell(cellId):
+    return array('d',(x(cellId),y(cellId),z(cellId)))
+
+
+from ROOT import gSystem, THnSparseD
```
Comment (hegner) [inline DetectorDescription/DetSensitive/tests/scripts/cmpTrackerSdDetailed.py]: imports to the top!

File: DetectorDescription/DetSensitive/tests/scripts/cmpTrackerSdDetailed.py [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+from EventStore import EventStore
+from array import array
+
```
Comment (hegner) [inline DetectorDescription/DetSensitive/tests/scripts/cmpTrackerSdDetailed.py]: I always prefer `if __name__ == "__main__":` etc if a script as well contains definitions of functions and methods.

File: DetectorDescription/DetSensitive/tests/scripts/cmpTrackerSdDetailed.py [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+for iev, event in enumerate(store2):
+    hits = event.get('hits')
+    for hit in hits:
+        cellId = hit.Core().Cellid
+        hE = hit.Core().Energy
+        hist2.Fill(cell(cellId),hE)
+for ix in range(cellNo):
+    for iy in range(cellNo):
+        for iz in range(cellNo):
+            assert(hist1.GetBinContent(array('i',(ix,iy,iz))) == hist2.GetBinContent(array('i',(ix,iy,iz))))
```
Comment (hegner) [inline DetectorDescription/DetSensitive/tests/scripts/cmpTrackerSdDetailed.py]: we eventually could use the unit test capabilities of python to make it somewhat nicer

**Comment by hegner** [comment]
Very nice! Just some cosmetic changes like proper python scripts, nullptr and clarifying ownership would be good.

**Comment by jlingema** [comment]
retest this please

**Comment by jlingema** [comment]
retest this please

**Pull request merged**